File: docs/eos/examples/compliance-starter/Redaction.java
// Redaction.java — strip regulated fields BEFORE any third-party / cross-border / LLM call.
// Zero-dependency (JDK only), deterministic, offline. Java parallel of redaction.mjs / redaction.py.
// This is the *code form* of the Agentic data-boundary (docs/checklists/F-compliance.md) and
// eos-doctor D5, enforcing "Never place secrets or PII in prompts or logs. Redact before sending
// to the provider." (.github/instructions/ai/10-ai-llm.instructions.md).
//
// Regime presets (compose only what your /compliance profile selected):
//   Redaction.createRedactor(List.of("PCI-DSS"))       // base + cardholder data only
//   Redaction.createRedactor(List.of("HIPAA"))         // base + PHI (identifiers + health fields)
//   Redaction.createRedactor(List.of("GDPR", "PIPL"))  // base + personal data
//   Redaction.createRedactor()                         // all regimes (also the static default below)
//   Redaction.createRedactor(List.of())                // base only (credentials/secrets)
//
// The values walked are decoded-JSON shapes: Map<String,Object>, List<Object>, String, scalars.
// A Redactor exposes redact / scan / assertClean.
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringJoiner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Redaction {

    // ── Luhn (card) check so we only redact real PANs, not any 13–19 digit id ──────────
    private static final Pattern DIGITS_13_19 = Pattern.compile("^\\d{13,19}$");

    public static boolean luhnValid(String digits) {
        if (digits == null || !DIGITS_13_19.matcher(digits).matches()) {
            return false;
        }
        int sum = 0;
        boolean alt = false;
        for (int i = digits.length() - 1; i >= 0; i--) {
            int n = digits.charAt(i) - '0';
            if (alt) {
                n *= 2;
                if (n > 9) {
                    n -= 9;
                }
            }
            sum += n;
            alt = !alt;
        }
        return sum % 10 == 0;
    }

    // ── Value patterns (catch PII in free text / unknown keys). validate gates a match. ─
    @FunctionalInterface
    interface Validator {
        boolean ok(String match);
    }

    static final class Pat {
        final String label;
        final Pattern re;
        final Validator validate; // null == always mask on match

        Pat(String label, String regex, Validator validate) {
            this.label = label;
            this.re = Pattern.compile(regex);
            this.validate = validate;
        }
    }

    private static final Pattern NON_DIGIT = Pattern.compile("\\D");
    static final Pat EMAIL = new Pat("email", "[\\w.+-]+@[\\w-]+\\.[\\w.-]+", null);
    static final Pat SSN = new Pat("ssn", "\\b\\d{3}-\\d{2}-\\d{4}\\b", null);
    static final Pat PHONE = new Pat("phone", "\\b\\d{3}[-.\\s]?\\d{3}[-.\\s]?\\d{4}\\b", null);
    static final Pat PAN = new Pat("pan", "\\b(?:\\d[ -]?){13,19}\\b",
            m -> luhnValid(NON_DIGIT.matcher(m).replaceAll("")));

    // Shared "direct personal identifiers" — HIPAA (18 Safe Harbor items) and GDPR/PIPL both want these out.
    static final List<String> PERSONAL_KEYS = List.of("email", "phone", "telephone", "fax", "dob",
            "dateofbirth", "birthdate", "address", "ssn", "socialsecurity");
    static final List<Pat> PERSONAL_PATTERNS = List.of(EMAIL, SSN, PHONE);

    static final class Profile {
        final List<String> keys;
        final List<Pat> patterns;

        Profile(List<String> keys, List<Pat> patterns) {
            this.keys = keys;
            this.patterns = patterns;
        }
    }

    private static List<String> withPersonal(String... extra) {
        List<String> out = new ArrayList<>(PERSONAL_KEYS);
        Collections.addAll(out, extra);
        return out;
    }

    // ── Regime profiles. keys = case-insensitive SUBSTRING match on object keys. ─────────
    // ⚠ Substring matching means SHORT tokens over-match ('pin'⊂'shipping', 'track'⊂'tracking',
    //   'sin'⊂'business'). Tokens below are curated to be distinctive; if your schema still
    //   collides, switch isDeniedKey to exact-match or scope the key list per module.
    public static final Map<String, Profile> PROFILES = buildProfiles();

    private static Map<String, Profile> buildProfiles() {
        Map<String, Profile> m = new LinkedHashMap<>();
        // Credentials/secrets — never leave the boundary regardless of regulatory regime.
        m.put("base", new Profile(List.of("password", "passwd", "secret", "token", "apikey",
                "api_key", "accesskey", "credential", "privatekey", "private_key"), List.of()));
        // HIPAA: PHI = personal identifiers + health-specific fields (Security/Privacy Rule, Safe Harbor).
        m.put("HIPAA", new Profile(withPersonal("mrn", "medicalrecord", "medical_record", "phi",
                "diagnosis", "icd", "healthplan", "health_plan", "beneficiary", "npi", "dea",
                "patientid"), PERSONAL_PATTERNS));
        // PCI-DSS: cardholder data (CHD) + sensitive authentication data (SAD) only — scoped to cards.
        m.put("PCI", new Profile(List.of("pan", "cardnumber", "card_number", "cardno", "cvv",
                "cvc", "cvv2", "track1", "track2", "magstripe", "pinblock", "expiry", "expiration"),
                List.of(PAN)));
        // GDPR / CCPA / PIPL: personal data — identifiers + national/sensitive PI.
        m.put("GDPR_PIPL", new Profile(withPersonal("nationalid", "national_id", "idcard",
                "passport", "taxid", "drivinglicense", "driverslicense", "biometric", "geolocation"),
                PERSONAL_PATTERNS));
        return Collections.unmodifiableMap(m);
    }

    // Map the regime names /compliance uses onto profile keys.
    static final Map<String, String> REGIME_ALIASES = Map.ofEntries(
            Map.entry("hipaa", "HIPAA"),
            Map.entry("pci", "PCI"), Map.entry("pci-dss", "PCI"), Map.entry("pcidss", "PCI"),
            Map.entry("gdpr", "GDPR_PIPL"), Map.entry("ccpa", "GDPR_PIPL"),
            Map.entry("ccpa-cpra", "GDPR_PIPL"), Map.entry("pipl", "GDPR_PIPL"));

    static final List<String> DEFAULT_REGIMES = List.of("HIPAA", "PCI", "GDPR_PIPL");
    private static final Pattern MASK_RE = Pattern.compile("^«REDACTED:[^»]+»$");

    static String mask(String label) {
        return "«REDACTED:" + label + "»";
    }

    static boolean isMask(Object v) {
        return v instanceof String s && MASK_RE.matcher(s).matches();
    }

    static boolean isEmpty(Object v) {
        return v == null || "".equals(v);
    }

    static String toProfileName(String r) {
        if (PROFILES.containsKey(r)) {
            return r;
        }
        return REGIME_ALIASES.get(r.toLowerCase());
    }

    /** One leak location reported by scan (path + kind). */
    public static final class Hit {
        public final String path;
        public final String kind;

        Hit(String path, String kind) {
            this.path = path;
            this.kind = kind;
        }

        @Override
        public String toString() {
            return path + " (" + kind + ")";
        }
    }

    /** Redactor bound to a set of regimes. Build via createRedactor. */
    public static final class Redactor {
        public final List<String> regimes; // active regimes (base excluded)
        public final List<String> keys;
        public final List<String> patternLabels;
        private final List<Pat> patterns;

        Redactor(List<String> regimes) {
            List<String> names = new ArrayList<>();
            names.add("base");
            List<String> src = (regimes != null) ? regimes : DEFAULT_REGIMES;
            for (String r : src) {
                String n = toProfileName(r);
                if (n != null && !names.contains(n)) {
                    names.add(n);
                }
            }
            List<String> ks = new ArrayList<>();
            List<Pat> ps = new ArrayList<>();
            Set<String> seenKey = new HashSet<>();
            Set<String> seenPat = new HashSet<>();
            for (String n : names) {
                Profile prof = PROFILES.get(n);
                if (prof == null) {
                    continue;
                }
                for (String k : prof.keys) {
                    String lk = k.toLowerCase();
                    if (seenKey.add(lk)) {
                        ks.add(lk);
                    }
                }
                for (Pat p : prof.patterns) {
                    if (seenPat.add(p.label)) {
                        ps.add(p);
                    }
                }
            }
            // Validate-gated patterns (PAN) first — they consume digit runs before generic matchers.
            ps.sort((a, b) -> Integer.compare(b.validate != null ? 1 : 0, a.validate != null ? 1 : 0));
            List<String> active = new ArrayList<>();
            for (String n : names) {
                if (!n.equals("base")) {
                    active.add(n);
                }
            }
            List<String> labels = new ArrayList<>();
            for (Pat p : ps) {
                labels.add(p.label);
            }
            this.regimes = Collections.unmodifiableList(active);
            this.keys = Collections.unmodifiableList(ks);
            this.patternLabels = Collections.unmodifiableList(labels);
            this.patterns = ps;
        }

        private boolean isDeniedKey(String k) {
            String lk = k.toLowerCase();
            for (String d : keys) {
                if (lk.contains(d)) {
                    return true;
                }
            }
            return false;
        }

        private String redactString(String s) {
            String out = s;
            for (Pat p : patterns) {
                Matcher mt = p.re.matcher(out);
                StringBuilder sb = new StringBuilder();
                while (mt.find()) {
                    String m = mt.group();
                    String rep = (p.validate == null || p.validate.ok(m)) ? mask(p.label) : m;
                    mt.appendReplacement(sb, Matcher.quoteReplacement(rep));
                }
                mt.appendTail(sb);
                out = sb.toString();
            }
            return out;
        }

        /** Deep copy with regulated fields masked (safe to send / log). */
        public Object redact(Object value) {
            if (value instanceof List<?> list) {
                List<Object> out = new ArrayList<>(list.size());
                for (Object x : list) {
                    out.add(redact(x));
                }
                return out;
            }
            if (value instanceof Map<?, ?> map) {
                Map<String, Object> out = new LinkedHashMap<>();
                for (Map.Entry<?, ?> e : map.entrySet()) {
                    String k = String.valueOf(e.getKey());
                    out.put(k, isDeniedKey(k) ? mask(k.toLowerCase()) : redact(e.getValue()));
                }
                return out;
            }
            if (value instanceof String s) {
                return redactString(s);
            }
            return value;
        }

        /** PII-free logging alias for redact. */
        public Object redactForLog(Object value) {
            return redact(value);
        }

        /** Report every place regulated data would leak (path + kind). Empty == clean. */
        public List<Hit> scan(Object value) {
            List<Hit> hits = new ArrayList<>();
            walk(value, "$", hits);
            return hits;
        }

        private void walk(Object v, String p, List<Hit> hits) {
            if (v instanceof List<?> list) {
                for (int i = 0; i < list.size(); i++) {
                    walk(list.get(i), p + "[" + i + "]", hits);
                }
            } else if (v instanceof Map<?, ?> map) {
                for (Map.Entry<?, ?> e : map.entrySet()) {
                    String k = String.valueOf(e.getKey());
                    Object val = e.getValue();
                    if (isDeniedKey(k) && !isEmpty(val) && !isMask(val)) {
                        hits.add(new Hit(p + "." + k, "key:" + k.toLowerCase()));
                    }
                    walk(val, p + "." + k, hits);
                }
            } else if (v instanceof String s) {
                if (!redactString(s).equals(s)) {
                    hits.add(new Hit(p, "value-pattern"));
                }
            }
        }

        /** Boundary guard — call right before any cross-border / third-party / LLM request.
         *  Throws IllegalStateException (Java's idiom) if regulated data would leave the boundary. */
        public Object assertClean(Object value) {
            List<Hit> hits = scan(value);
            if (!hits.isEmpty()) {
                StringJoiner sj = new StringJoiner(", ");
                for (Hit h : hits) {
                    sj.add(h.toString());
                }
                throw new IllegalStateException(
                        "redaction: " + hits.size() + " regulated field(s) would leave the boundary: " + sj);
            }
            return value;
        }
    }

    /** Build a redactor for all regimes (the default). */
    public static Redactor createRedactor() {
        return new Redactor(null);
    }

    /** Build a redactor for the selected regimes (names/aliases; empty list == base only). */
    public static Redactor createRedactor(List<String> regimes) {
        return new Redactor(regimes);
    }

    // ── Auto-select the regime(s) from the /compliance output ──────────────────────────
    // `/compliance` writes docs/compliance-profile.md with a canonical machine-readable line:
    //   **Regulatory regime:** HIPAA, PCI-DSS      (or `none` for generic PII handling)
    private static final Pattern REGIME_LINE =
            Pattern.compile("(?i)regulatory\\s+regime\\s*[:\uFF1A]\\s*(.+)");
    private static final Pattern REGIME_SPLIT = Pattern.compile("[\\s,/+&]+");
    private static final Pattern NONE_PREFIX = Pattern.compile("(?i)^none\\b");

    /**
     * Extract resolved regime names from the profile's "Regulatory regime:" line. Only tokens
     * that resolve to a known regime are kept, so free-text rationale is ignored; an explicit
     * {@code none} (or no match) yields an empty list, which createRedactor scopes to base
     * credentials only.
     */
    public static List<String> parseRegimes(String profileText) {
        List<String> out = new ArrayList<>();
        if (profileText == null) {
            return out;
        }
        Matcher m = REGIME_LINE.matcher(profileText);
        if (!m.find()) {
            return out;
        }
        String val = m.group(1).replaceFirst("^[*_\\s]+", ""); // drop leading **bold**
        for (String p : new String[] {"(", "\uFF08"}) {        // drop parenthetical rationale
            int i = val.indexOf(p);
            if (i >= 0) {
                val = val.substring(0, i);
            }
        }
        val = val.trim();
        if (NONE_PREFIX.matcher(val).find()) {
            return out;
        }
        for (String tok : REGIME_SPLIT.split(val)) {
            if (tok.isEmpty()) {
                continue;
            }
            String n = toProfileName(tok);
            if (n != null && !n.equals("base") && !out.contains(n)) {
                out.add(n);
            }
        }
        return out;
    }

    /** Build a redactor from the /compliance profile doc (default fallback: all regimes). */
    public static Redactor redactorFromProfile(String path) {
        return redactorFromProfile(path, "all");
    }

    /**
     * Build a redactor straight from the /compliance profile doc.
     * <ul>
     *   <li>file lists regimes -&gt; scope to them
     *   <li>file says {@code none} -&gt; base credentials only (honours the human decision)
     *   <li>file missing -&gt; {@code fallback}: {@code "all"} (default, fail-safe — never
     *       silently under-redact), {@code "base"}, or {@code "throw"}.
     * </ul>
     */
    public static Redactor redactorFromProfile(String path, String fallback) {
        String p = (path != null && !path.isEmpty()) ? path : "docs/compliance-profile.md";
        String text;
        try {
            text = Files.readString(Path.of(p));
        } catch (IOException e) {
            if ("throw".equals(fallback)) {
                throw new UncheckedIOException(
                        "redactorFromProfile: " + p + " not found — run /compliance first", e);
            }
            return createRedactor("base".equals(fallback) ? List.of() : null);
        }
        return createRedactor(parseRegimes(text));
    }

    // ── Default instance = all regimes. Static convenience API. ────────────────────────
    private static final Redactor DEFAULT = createRedactor();

    public static Object redact(Object v) {
        return DEFAULT.redact(v);
    }

    public static Object redactForLog(Object v) {
        return DEFAULT.redactForLog(v);
    }

    public static List<Hit> scan(Object v) {
        return DEFAULT.scan(v);
    }

    public static Object assertClean(Object v) {
        return DEFAULT.assertClean(v);
    }

    /** Composed deny list of the default (all-regime) instance — for inspection / back-compat. */
    public static final List<String> DENY_KEYS = DEFAULT.keys;

    private Redaction() {
    }
}

File: docs/eos/examples/compliance-starter/redaction.go
// redaction.go — strip regulated fields BEFORE any third-party / cross-border / LLM call.
// Zero-dependency (stdlib only), deterministic, offline. Go parallel of redaction.mjs / redaction.py.
// This is the *code form* of the Agentic data-boundary (docs/checklists/F-compliance.md) and
// eos-doctor D5, enforcing "Never place secrets or PII in prompts or logs. Redact before sending
// to the provider." (.github/instructions/ai/10-ai-llm.instructions.md).
//
// Regime presets (compose only what your /compliance profile selected):
//
//	CreateRedactor([]string{"PCI-DSS"})       // base + cardholder data only
//	CreateRedactor([]string{"HIPAA"})         // base + PHI (personal identifiers + health fields)
//	CreateRedactor([]string{"GDPR", "PIPL"})  // base + personal data
//	CreateRedactor(nil)                       // all regimes (also the package-level default below)
//	CreateRedactor([]string{})                // base only (credentials/secrets)
//
// The values walked are the decoded-JSON shapes: map[string]any, []any, string, and scalars.
// A Redactor exposes Redact / Scan / AssertClean.
package compliance

import (
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
)

// ── Luhn (card) check so we only redact real PANs, not any 13–19 digit id ──────────────
var reDigits1319 = regexp.MustCompile(`^\d{13,19}$`)

// LuhnValid reports whether digits (13–19 chars) pass the Luhn checksum.
func LuhnValid(digits string) bool {
	if !reDigits1319.MatchString(digits) {
		return false
	}
	sum, alt := 0, false
	for i := len(digits) - 1; i >= 0; i-- {
		n := int(digits[i] - '0')
		if alt {
			n *= 2
			if n > 9 {
				n -= 9
			}
		}
		sum += n
		alt = !alt
	}
	return sum%10 == 0
}

// ── Value patterns (catch PII in free text / unknown keys). validate gates a match. ────
type pattern struct {
	label    string
	re       *regexp.Regexp
	validate func(string) bool // nil == always mask on match
}

var nonDigit = regexp.MustCompile(`\D`)

var (
	patEmail = pattern{"email", regexp.MustCompile(`[\w.+-]+@[\w-]+\.[\w.-]+`), nil}
	patSSN   = pattern{"ssn", regexp.MustCompile(`\b\d{3}-\d{2}-\d{4}\b`), nil}
	patPhone = pattern{"phone", regexp.MustCompile(`\b\d{3}[-.\s]?\d{3}[-.\s]?\d{4}\b`), nil}
	patPAN   = pattern{"pan", regexp.MustCompile(`\b(?:\d[ -]?){13,19}\b`), func(m string) bool { return LuhnValid(nonDigit.ReplaceAllString(m, "")) }}
)

// Shared "direct personal identifiers" — HIPAA (18 Safe Harbor items) and GDPR/PIPL both want these out.
var personalKeys = []string{"email", "phone", "telephone", "fax", "dob", "dateofbirth", "birthdate", "address", "ssn", "socialsecurity"}
var personalPatterns = []pattern{patEmail, patSSN, patPhone}

// Profile is one regulatory regime's deny config. Fields are unexported; use the API below.
type Profile struct {
	keys     []string
	patterns []pattern
}

// withPersonal returns a copy of personalKeys plus extras (never aliases the shared slice).
func withPersonal(extra ...string) []string {
	return append(append([]string{}, personalKeys...), extra...)
}

// ── Regime profiles. keys = case-insensitive SUBSTRING match on object keys. ───────────
// ⚠ Substring matching means SHORT tokens over-match ('pin'⊂'shipping', 'track'⊂'tracking',
//
//	'sin'⊂'business'). Tokens below are curated to be distinctive; if your schema still
//	collides, switch isDeniedKey to exact-match or scope the key list per module.
//
// Profiles holds every regime preset, keyed by profile name.
var Profiles = map[string]Profile{
	// Credentials/secrets — never leave the boundary regardless of regulatory regime.
	"base": {keys: []string{"password", "passwd", "secret", "token", "apikey", "api_key", "accesskey", "credential", "privatekey", "private_key"}, patterns: nil},
	// HIPAA: PHI = personal identifiers + health-specific fields (Security/Privacy Rule, Safe Harbor).
	"HIPAA": {keys: withPersonal("mrn", "medicalrecord", "medical_record", "phi", "diagnosis", "icd", "healthplan", "health_plan", "beneficiary", "npi", "dea", "patientid"), patterns: personalPatterns},
	// PCI-DSS: cardholder data (CHD) + sensitive authentication data (SAD) only — scoped to cards.
	"PCI": {keys: []string{"pan", "cardnumber", "card_number", "cardno", "cvv", "cvc", "cvv2", "track1", "track2", "magstripe", "pinblock", "expiry", "expiration"}, patterns: []pattern{patPAN}},
	// GDPR / CCPA / PIPL: personal data — identifiers + national/sensitive PI.
	"GDPR_PIPL": {keys: withPersonal("nationalid", "national_id", "idcard", "passport", "taxid", "drivinglicense", "driverslicense", "biometric", "geolocation"), patterns: personalPatterns},
}

// RegimeAliases maps the regime names /compliance uses onto profile keys.
var RegimeAliases = map[string]string{
	"hipaa": "HIPAA",
	"pci":   "PCI", "pci-dss": "PCI", "pcidss": "PCI",
	"gdpr": "GDPR_PIPL", "ccpa": "GDPR_PIPL", "ccpa-cpra": "GDPR_PIPL", "pipl": "GDPR_PIPL",
}

var defaultRegimes = []string{"HIPAA", "PCI", "GDPR_PIPL"}
var reMask = regexp.MustCompile(`^«REDACTED:[^»]+»$`)

func mask(label string) string { return "«REDACTED:" + label + "»" }

func isMask(v any) bool {
	s, ok := v.(string)
	return ok && reMask.MatchString(s)
}

func isEmpty(v any) bool {
	if v == nil {
		return true
	}
	s, ok := v.(string)
	return ok && s == ""
}

func toProfileName(r string) string {
	if _, ok := Profiles[r]; ok {
		return r
	}
	return RegimeAliases[strings.ToLower(r)]
}

// Hit is one leak location reported by Scan (path + kind).
type Hit struct {
	Path string
	Kind string
}

// Redactor is bound to a set of regimes. Build via CreateRedactor.
type Redactor struct {
	Regimes  []string // active regimes (base excluded)
	keys     []string
	patterns []pattern
}

// CreateRedactor builds a redactor for the selected regimes. Pass nil for all regimes
// (the default), an empty slice for base-only, or names/aliases like "HIPAA" / "pci-dss".
func CreateRedactor(regimes []string) *Redactor {
	names := []string{"base"}
	src := regimes
	if regimes == nil {
		src = defaultRegimes
	}
	for _, r := range src {
		if n := toProfileName(r); n != "" && !contains(names, n) {
			names = append(names, n)
		}
	}
	var keys []string
	var pats []pattern
	seenKey := map[string]bool{}
	seenPat := map[string]bool{}
	for _, n := range names {
		p, ok := Profiles[n]
		if !ok {
			continue
		}
		for _, k := range p.keys {
			lk := strings.ToLower(k)
			if !seenKey[lk] {
				seenKey[lk] = true
				keys = append(keys, lk)
			}
		}
		for _, pt := range p.patterns {
			if !seenPat[pt.label] {
				seenPat[pt.label] = true
				pats = append(pats, pt)
			}
		}
	}
	// Validate-gated patterns (PAN) first — they consume digit runs before generic matchers.
	sort.SliceStable(pats, func(i, j int) bool {
		return pats[i].validate != nil && pats[j].validate == nil
	})
	active := []string{}
	for _, n := range names {
		if n != "base" {
			active = append(active, n)
		}
	}
	return &Redactor{Regimes: active, keys: keys, patterns: pats}
}

func contains(xs []string, x string) bool {
	for _, v := range xs {
		if v == x {
			return true
		}
	}
	return false
}

// Keys returns the composed deny-key list (lowercased, deduped).
func (r *Redactor) Keys() []string { return append([]string{}, r.keys...) }

// PatternLabels returns the active value-pattern labels in match order.
func (r *Redactor) PatternLabels() []string {
	out := make([]string, len(r.patterns))
	for i, p := range r.patterns {
		out[i] = p.label
	}
	return out
}

func (r *Redactor) isDeniedKey(k string) bool {
	lk := strings.ToLower(k)
	for _, d := range r.keys {
		if strings.Contains(lk, d) {
			return true
		}
	}
	return false
}

func (r *Redactor) redactString(s string) string {
	out := s
	for _, p := range r.patterns {
		p := p // capture per iteration (safe on every Go version)
		if p.validate == nil {
			out = p.re.ReplaceAllString(out, mask(p.label))
		} else {
			out = p.re.ReplaceAllStringFunc(out, func(m string) string {
				if p.validate(m) {
					return mask(p.label)
				}
				return m
			})
		}
	}
	return out
}

// Redact returns a deep copy with regulated fields masked (safe to send / log).
func (r *Redactor) Redact(value any) any {
	switch val := value.(type) {
	case []any:
		out := make([]any, len(val))
		for i, x := range val {
			out[i] = r.Redact(x)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(val))
		for k, v := range val {
			if r.isDeniedKey(k) {
				out[k] = mask(strings.ToLower(k))
			} else {
				out[k] = r.Redact(v)
			}
		}
		return out
	case string:
		return r.redactString(val)
	default:
		return value
	}
}

// RedactForLog is a PII-free logging alias for Redact.
func (r *Redactor) RedactForLog(value any) any { return r.Redact(value) }

// Scan reports every place regulated data would leak (path + kind). Empty == clean.
func (r *Redactor) Scan(value any) []Hit {
	var hits []Hit
	var walk func(v any, p string)
	walk = func(v any, p string) {
		switch val := v.(type) {
		case []any:
			for i, x := range val {
				walk(x, fmt.Sprintf("%s[%d]", p, i))
			}
		case map[string]any:
			for k, vv := range val {
				if r.isDeniedKey(k) && !isEmpty(vv) && !isMask(vv) {
					hits = append(hits, Hit{Path: p + "." + k, Kind: "key:" + strings.ToLower(k)})
				}
				walk(vv, p+"."+k)
			}
		case string:
			if r.redactString(val) != val {
				hits = append(hits, Hit{Path: p, Kind: "value-pattern"})
			}
		}
	}
	walk(value, "$")
	return hits
}

// AssertClean is the boundary guard — call it right before any cross-border / third-party /
// LLM request. It returns a non-nil error (Go's idiom instead of throwing) if regulated data
// would leave the boundary.
func (r *Redactor) AssertClean(value any) error {
	hits := r.Scan(value)
	if len(hits) == 0 {
		return nil
	}
	parts := make([]string, len(hits))
	for i, h := range hits {
		parts[i] = fmt.Sprintf("%s (%s)", h.Path, h.Kind)
	}
	return fmt.Errorf("redaction: %d regulated field(s) would leave the boundary: %s", len(hits), strings.Join(parts, ", "))
}

// ── Default instance = all regimes. Package-level convenience API. ─────────────────────
var defaultRedactor = CreateRedactor(nil)

var (
	reRegimeLine   = regexp.MustCompile(`(?i)regulatory\s+regime\s*[:：]\s*(.+)`)
	reRegimeSplit  = regexp.MustCompile(`[\s,/+&]+`)
	reLeadEmphasis = regexp.MustCompile(`^[*_\s]+`)
	reNonePrefix   = regexp.MustCompile(`(?i)^none\b`)
)

// ParseRegimes extracts resolved regime names from the "Regulatory regime:" line that
// /compliance writes into docs/compliance-profile.md. Only tokens that resolve to a known
// regime are kept, so free-text rationale on the line is ignored; an explicit `none` (or no
// match) yields a non-nil empty slice, which CreateRedactor scopes to base credentials only.
func ParseRegimes(profileText string) []string {
	out := []string{}
	m := reRegimeLine.FindStringSubmatch(profileText)
	if m == nil {
		return out
	}
	val := reLeadEmphasis.ReplaceAllString(m[1], "") // drop leading **bold**
	for _, p := range []string{"(", "（"} {           // drop parenthetical rationale
		if i := strings.Index(val, p); i >= 0 {
			val = val[:i]
		}
	}
	val = strings.TrimSpace(val)
	if reNonePrefix.MatchString(val) {
		return out
	}
	for _, tok := range reRegimeSplit.Split(val, -1) {
		if tok == "" {
			continue
		}
		if n := toProfileName(tok); n != "" && n != "base" && !contains(out, n) {
			out = append(out, n)
		}
	}
	return out
}

// RedactorFromProfile builds a redactor from the /compliance profile doc. Pass "" to use the
// default path (docs/compliance-profile.md). A present file scopes to its regimes (or base
// only for `none`); a missing file returns the fail-safe all-regime redactor together with the
// read error, so a caller may surface it or ignore it and keep the strict default — never
// silently under-redacting.
func RedactorFromProfile(path string) (*Redactor, error) {
	if path == "" {
		path = "docs/compliance-profile.md"
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return CreateRedactor(nil), err
	}
	return CreateRedactor(ParseRegimes(string(data))), nil
}

// Redact masks regulated fields using the all-regime default instance.
func Redact(v any) any { return defaultRedactor.Redact(v) }

// RedactForLog is the PII-free logging alias on the default instance.
func RedactForLog(v any) any { return defaultRedactor.RedactForLog(v) }

// Scan reports leaks using the default instance.
func Scan(v any) []Hit { return defaultRedactor.Scan(v) }

// AssertClean is the boundary guard on the default instance.
func AssertClean(v any) error { return defaultRedactor.AssertClean(v) }

// DenyKeys is the composed deny list of the default (all-regime) instance — for inspection.
var DenyKeys = defaultRedactor.keys
